frontend/utils/distribution/group-income-distribution.js

```diff
@@ -1,40 +1,253 @@
+    .filter(([,profile]) => profile.joinedDate.startsWith(monthstamp))
+    .map(mapUser({ withDate: true }))
+    .map(user => { user.type = 'join'; return user }))
+
+  const events = (paymentEvents
+    .concat(joinEvents)
+    .sort((a, b) => a.date < b.date ? -1 : 1)
+    .map(event => { delete event.date; return event }))
+
+  const members = (Object.entries(data.groupProfiles)
+    .filter(([,profile]) => profile.joinedDate < monthstamp)
+    .map(mapUser({ withDate: false })))
+
+  return {
+    haves: members.filter(user => user.have !== undefined && user.have > 0),
+    needs: members.filter(user => user.need !== undefined),
+    events
+  }
+}
+
+/*
+JS is giving us 20 * (10 / 18) * (18 / 50) = 3.9999999999999996 instead of 4
+*/
+
+function distibuteFromHavesToNeeds({ haves, needs }) {
+  const totalHave = haves.reduce((a, b) => a + b.have, 0)
+  const totalNeed = needs.reduce((a, b) => a + b.need, 0)
+  const totalPercent = Math.min(1, totalHave / totalNeed)
+
+  for (const have of haves) have.percent = have.have / totalHave
+
+  const payments = []
+  for (const need of needs) {
+    for (const have of haves) {
+      const amount = need.need * have.percent * totalPercent
+      need.need -= amount
+      payments.push({ amount, from: have.name, to: need.name })
     }
   }
-  var dist = incomeDistribution(currentIncomeDistribution, mincomeAmount)
-  if (adjusted) {
+  return payments
+}
+
+function handlePayment({ payment, payments, needs }) {
+  const { from, to, amount } = payment
+
+  for (const payment of payments) {
+    if (payment.from === from && payment.to === to) {
+      payment.amount -= amount
+      if (payment.amount < 0) {
+        const totalOverPayment = -payment.amount
+        payment.amount = 0
+
+        const myOtherPayments = payments.filter(otherPayment =>
+          otherPayment !== payment && otherPayment.from === from
+        )
+        for (const otherPayment of myOtherPayments) {
+          const overPayment = (totalOverPayment / myOtherPayments.length)
+          const otherTo = needs.find(u => u.name === otherPayment.to)
+          otherPayment.amount -= overPayment
+          otherTo.need += overPayment
+        }
+      }
+    }
+  }
+}
+
+function handleJoin({ user, needs }) {
+  const { name, have, need } = user
+  if (have !== undefined) {
+    return distibuteFromHavesToNeeds({ needs, haves: [{ name, have }] })
+  }
+  else {
+    // TODO: This scenario doesn't have tests yet
+    // needs.push({ name, need })
+  }
+}
+
+// Note: this mutates the objects in haves/needs
+export function groupIncomeDistributionNewLogic ({ haves, needs, events }) {
+  let payments = distibuteFromHavesToNeeds({ haves, needs })
+
+  for (const event of events) {
+    if (event.type === 'payment') {
+      handlePayment({ payments, needs, payment: event })
+    }
+    else if (event.type === 'join') {
+      const newPayments = handleJoin({ user: event, needs })
+      payments = payments.concat(newPayments)
+    }
```

Comment:
Minor: I have made it a habit to recognize if/else blocks like this, and write them in a way that is slightly different. Since if/else always creates a branch, and since branches slow things down (and can also make code slightly less readable in some situations), I use a map in situations like this:

```js
for (const event of events) {
  ({
    payment: event => handlePayment({ payments, needs, payment: event }),
    join: event => {
      const newPayments = handleJoin({ user: event, needs })
      payments = payments.concat(newPayments)
    }
  })[event.type](event)
}
```

It also has the added benefit of automatically handling unknown types by throwing an exception.

Reply:
Thanks for letting me know. If you mean it slows things down in the JavaScript VM, looped function creation and execution is probably much slower than if-branching, although performance doesn't really matter here.

Reply:
Oh yeah, I don't know if the JS compiler would actually do that, but if it did then yeah it would make sense to move the definition of that object outside of the loop.